### PR TITLE
Generate all variations complete flow

### LINKED ISSUE #7877: Generate all variations when user requested it
### Description

This task is about wrapping up the complete user journey for the "Generate all variations" feature.

After finishing:
- #7856
- #7857
- #7859
- #7858
- #7886 

we have to connect this work in one flow.

### PR DESCRIPTION
**Pull request by wzieba**
Closes: #7877

### Description

This PR replaces mocked behavior with a real implementation of the API call for creating generated variations.

It also improves the flow and adds unit tests to the whole implementation.

#### TC1 Generate variations
1. Execute this on a product with not a great number of attributes. 2 or 3 attributes with 2 options would be just fine.
1. Open details of a variable product that has at least one variation
2. Tap on `Variations`
3. Tap on `Generate all variations`
4. **Assert there's correct number of variations to be generated in dialog's message**
5. Tap on `ok`
6. **Assert progress bar is visible**
7. **Assert that after a second, progress bar disappeared and you can see generated variations**

#### TC2 Attempt to generate more than 100 variations
1. Execute this on a product with a great number of attributes. E.g. 3 attributes with 5, 3 and 7 options will be ok.
1. Open details of a variable product that has at least one variation 
2. Tap on `Variations`
3. Tap on `Generate all variations`
4. **Assert there's a message about exceeding number of variations to be generated**
4. **Assert that dialog disappears after tapping "ok" and there are no new variations**

#### TC3 Api failure
1. Open details of a variable product that has at least one variation 
2. Tap on `Variations`
3. Turn off the internet connection
4. Tap on `Generate all variations`
5. Confirm variations generateion
6. **Assert you can see a generic network error message**

### Images/gif

https://user-images.githubusercontent.com/5845095/203362061-86e9fdf1-d1e0-4d96-97b4-d7e3fe6909b7.mov

- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/variations/VariationListViewModel.kt [hunk 1]
```diff
@@ -70,7 +71,9 @@ class VariationListViewModel @Inject constructor(
     private val dispatchers: CoroutineDispatchers,
     private val generateVariationCandidates: GenerateVariationCandidates,
 ) : ScopedViewModel(savedState) {
-    private var remoteProductId = 0L
+
+    private val navArgs: VariationListFragmentArgs by savedState.navArgs()
+    private val remoteProductId = navArgs.remoteProductId
```
Comment (wzieba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/variations/VariationListViewModel.kt:76]: Little refactor so we won't have to worry about mutability of `remoteProductId`

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/variations/VariationListViewModel.kt [hunk 2]
```diff
@@ -211,15 +214,13 @@ class VariationListViewModel @Inject constructor(
...
-        this.remoteProductId = remoteProductId
-
         loadingJob = launch {
             viewState = viewState.copy(isLoadingMore = loadMore)
             if (!loadMore) {
                 // if this is the initial load, first get the product variations from the db and if there are any show
                 // them immediately, otherwise make sure the skeleton shows
                 val variationsInDb = variationRepository.getProductVariationList(remoteProductId)
-                if (variationsInDb.isNullOrEmpty()) {
+                if (variationsInDb.isEmpty()) {
```
Comment (wzieba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/variations/VariationListViewModel.kt:223]: As `VariationRepository` returns `List<ProductVariation>` here, there was no point to check if the list is `null`. We only used that in tests (tests were mocking impossible scenarios).

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/variations/VariationListViewModel.kt [hunk 3]
```diff
@@ -346,8 +354,8 @@ class VariationListViewModel @Inject constructor(
 
     data class ShowGenerateVariationConfirmation(val variationCandidates: List<VariationCandidate>) : Event()
 
-    sealed class ShowGenerateVariationError : Event() {
-        data class LimitExceeded(val variationCandidatesSize: Int) : ShowGenerateVariationError()
-        object NetworkError : ShowGenerateVariationError()
+    sealed class ShowGenerateVariationsError : Event() {
```
Comment (wzieba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/variations/VariationListViewModel.kt:357]: I only added `s` here :)

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/variations/domain/GenerateVariationCandidates.kt [hunk 4]
```diff
@@ -10,11 +10,14 @@ import javax.inject.Inject
...
 class GenerateVariationCandidates @Inject constructor(
-    val variationRepository: VariationRepository
+    private val variationRepository: VariationRepository
 ) {
 
     companion object {
         const val VARIATION_CREATION_LIMIT = 100
+
+        private val initialAccumulator
+            get() = listOf(emptyList<VariationCandidate>())
```
Comment (wzieba) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/variations/domain/GenerateVariationCandidates.kt:20]: I want to keep `GenerateVariationCandidates` stateless and emphasize that. I've extracted this value to a backing field getter to increase readability.

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/variations/VariationRepository.kt [hunk 5]
```diff
@@ -105,6 +109,33 @@ class VariationRepository @Inject constructor(
...
+            WCProductStore.BatchGenerateVariationsPayload(
+                selectedSite.get(),
+                remoteProductId,
+                variations = variationCandidates.map { candidate ->
+                    candidate.map { variantOption ->
+                        WCProductVariationModel.ProductVariantOption(
+                            id = variantOption.id,
+                            name = variantOption.name,
+                            option = variantOption.option
+                        )
```
Comment (atorresveiga) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/variations/VariationRepository.kt]: Here we can use the `asSourceModel()` from `VariantOption`
```suggestion
                        variantOption.asSourceModel()
```

File: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/variations/domain/GenerateVariationCandidates.kt [hunk 6]
```diff
@@ -43,7 +46,7 @@ class GenerateVariationCandidates @Inject constructor(
             termAssignmentsGroupedByAttribute
         ).minus(existingVariationsAsCandidates.toSet())
 
-        return if (variationCandidates.first().isEmpty()) {
+        return if (variationCandidates == initialAccumulator) {
```
Comment (atorresveiga) [inline WooCommerce/src/main/kotlin/com/woocommerce/android/ui/products/variations/domain/GenerateVariationCandidates.kt:49]: Nice!

File: WooCommerce/src/test/kotlin/com/woocommerce/android/ui/products/VariationListViewModelTest.kt [hunk 7]
```diff
@@ -114,9 +134,141 @@ class VariationListViewModelTest : BaseUnitTest() {
...
+            .let { lastEvent ->
+                assertThat(lastEvent).isEqualTo(ShowGenerateVariationConfirmation(variationCandidates))
+            }
+    }
+
+    @Test
+    fun `Display limit exceeded error if user requested variations generation above the limit`() {
+        // given
+        val variationCandidatesAboveLimit =
+            (0..GenerateVariationCandidates.VARIATION_CREATION_LIMIT).map {
```
Comment (atorresveiga) [inline WooCommerce/src/test/kotlin/com/woocommerce/android/ui/products/VariationListViewModelTest.kt]: np: Nothing that needs to be changed, but here we can use the `List` class [initializer](https://kotlinlang.org/docs/constructing-collections.html#initializer-functions-for-lists) to achieve the same result:
```
List(GenerateVariationCandidates.VARIATION_CREATION_LIMIT) { 
         listOf(VariantOption(1, "Size", "S"))
}
```

**Review by atorresveiga — APPROVED** [review]
Great work @wzieba! The code looks good and everything is working as expected. I just left one little comment about using the `asSourceModel()` function, and a np that you can feel free to ignore. Both are minor changes, so I'm approving this PR.
LGTM! :shipit:

**Review by atorresveiga — CHANGES_REQUESTED** [review]
Making a second round of tests, after my initial review, I found one issue.

When a product has all variations generated. If I press generate all variations again, the app will fail with a network error (I believe we can improve this notice), but the most important is that we are sending an empty request and we should not.

https://user-images.githubusercontent.com/18119390/203453911-78fb3a36-f3e9-4d93-8999-8f3f861ec9a3.mp4

Screen Shot 2022-11-22 at 22 29 16

Do you think we should tackle this as part of this PR. I mean improving the notice. We can work on a different PR on FluxC throwing and exception when the list is empty. WDYT?

**Comment by wzieba** [comment]
I think we can work on this in a different PR, I've already created an issue for that: https://github.com/woocommerce/woocommerce-android/issues/7919

Adding a fix FluxC would also be a good idea!

File: WooCommerce/src/test/kotlin/com/woocommerce/android/ui/products/VariationListViewModelTest.kt (hunk 7 above)
Comment (wzieba) [reply WooCommerce/src/test/kotlin/com/woocommerce/android/ui/products/VariationListViewModelTest.kt]: Thanks, TIL!

**Review by atorresveiga — APPROVED** [review]
Great work @wzieba! I think that we are in a good place to merge these changes.
LGTM! :shipit:

**Pull request merged**